### PR TITLE
Add crypto_sign and crypto_sign_open to Signature

### PR DESCRIPTION
**Pull request by sporkins**

### CODE REVIEW

File: crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java [hunk 1]
```diff
@@ -502,4 +526,49 @@ public static boolean verifyDetached(byte[] message, byte[] signature, Signature
 
     return true;
   }
+
+  public static Bytes sign(Bytes message, Signature.SecretKey secretKey) {
```
Comment (atoulme) [inline crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: can you copy the javadoc from the method below here too please?

File: crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java [hunk 2]
```diff
@@ -502,4 +526,49 @@ public static boolean verifyDetached(byte[] message, byte[] signature, Signature
...
+    LongLongByReference signatureLengthReference = new LongLongByReference();
+    int rc = Sodium.crypto_sign(signature, signatureLengthReference, message, message.length, secretKey.bytesArray());
+    if (rc != 0) {
+      throw new SodiumException("crypto_sign: failed with result " + rc);
+    }
+
+    return signature;
+  }
+
+  public static Bytes openSigned(Bytes signed, Signature.PublicKey publicKey) {
```
Comment (atoulme) [inline crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: same

File: crypto/src/test/java/net/consensys/cava/crypto/sodium/SignatureTest.java [hunk 3]
```diff
@@ -43,4 +44,19 @@ void checkDetachedSignVerify() {
...
+  void checkSignAndVerify() {
+    Signature.KeyPair keyPair = Signature.KeyPair.random();
+    Bytes signed = Signature.sign(Bytes.fromHexString("deadbeef"), keyPair.secretKey());
+    Bytes messageBytes = Signature.openSigned(signed, keyPair.publicKey());
+    assertEquals(Bytes.fromHexString("deadbeef"), messageBytes);
+  }
+
+  @Test
+  void testSignatureSecretToBoxKeyPair() {
+    Signature.KeyPair signKeyPiar = Signature.KeyPair.random();
```
Comment (atoulme) [inline crypto/src/test/java/net/consensys/cava/crypto/sodium/SignatureTest.java]: typo here

File: crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java (hunk 1 above)
Comment (sporkins) [reply crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: javadoc added

File: crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java (hunk 2 above)
Comment (sporkins) [reply crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: javadoc added

File: crypto/src/test/java/net/consensys/cava/crypto/sodium/SignatureTest.java (hunk 3 above)
Comment (sporkins) [reply crypto/src/test/java/net/consensys/cava/crypto/sodium/SignatureTest.java]: typo fixed

File: crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java [hunk 4]
```diff
@@ -211,6 +211,21 @@ public Bytes sign(Bytes content) {
       return Signature.signDetached(content, this);
     }
 
+    /**
+     * converts signature key (Ed25519) to a box key (Curve25519) so that the same key pair can be used both for
+     * authenticated encryption and for signatures. See https://libsodium.gitbook.io/doc/advanced/ed25519-curve25519
+     *
+     * @return Box.KeyPair generated from Signature.KeyPair.SecretKey
+     */
+    public Box.KeyPair toBoxKeyPair() {
```
Comment (cleishm) [inline crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: As a matter of style, I'd prefer this as `Box.KeyPair.forSignatureKey(Signature.SecretKey key)`

Comment (sporkins) [reply crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: I thought about adding that as well.  Should I add there and remove from Signature.KeyPair?

File: crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java [hunk 5]
```diff
@@ -421,6 +436,15 @@ public KeyPair(Signature.PublicKey publicKey, Signature.SecretKey secretKey) {
       return secretKey;
     }
 
+    /**
+     * See Signature.SecretKey#toBoxKeyPair()
+     * 
+     * @return Box.KeyPair generated from Signature.KeyPair
+     */
+    public Box.KeyPair toBoxKeyPair() {
```
Comment (cleishm) [inline crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: Not sure we need both `Signature.SecretKey#toBoxKeyPair()` and `Signature.KeyPair#toBoxKeyPair()` - I'd make it _either_ `Box.KeyPair.fromSignatureSecretKey(Signature.SecretKey secretKey)` _or_ `Box.KeyPair.fromSignatureKeyPair(Signature.KeyPair keyPair)` with my preference being the latter.

File: crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java (hunk 4 above)
Comment (cleishm) [reply crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: That would be my preference. E.g. 297f00e1f03586058e4267c0fbbfcc70fe81fcb1

Comment (cleishm) [reply crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: Correction - d20454975f15d27d7dcf555dd8210b34cb2597ff

File: crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java [hunk 6]
```diff
@@ -502,4 +526,63 @@ public static boolean verifyDetached(byte[] message, byte[] signature, Signature
...
   }
+
+  /**
+   * Signs a message for a given key.
+   *
+   * @param message The message to sign.
+   * @param secretKey The secret key to sign the message with.
+   * @return The signature prepended to the message
+   */
+  public static Bytes sign(Bytes message, Signature.SecretKey secretKey) {
```
Comment (cleishm) [inline crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: No need to fully qualify `SecretKey` here.

File: crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java [hunk 7]
```diff
@@ -502,4 +526,63 @@ public static boolean verifyDetached(byte[] message, byte[] signature, Signature
...
+  }
+
+  /**
+   * Verifies the signature of the signed message using the public key and returns the message.
+   * 
+   * @param signed signed message (signature + message)
+   * @param publicKey pk used to verify the signature
+   * @return the message
+   */
+  public static Bytes openSigned(Bytes signed, Signature.PublicKey publicKey) {
```
Comment (cleishm) [inline crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: `openSigned` seems like the wrong name here. The existing methods are `signDetached` and `verifyDetached`. Should these be simply `sign` and `verify`?

File: crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java (hunk 4 above)
Comment (sporkins) [reply crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: Thanks for the suggestion, this is much better.  Moved this to Box.KeyPair and renamed method.

File: crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java (hunk 5 above)
Comment (sporkins) [reply crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: Agreed, kept only `Box.KeyPair.fromSignatureKeyPair(Signature.KeyPair keyPair)`

File: crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java (hunk 6 above)
Comment (sporkins) [reply crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: The other methods that were there already have it fully qualified.  I can change it if you want though.

File: crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java (hunk 7 above)
Comment (sporkins) [reply crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: Yes, `verify` is better and more readable.  Changed.

File: crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java (hunk 6 above)
Comment (cleishm) [reply crypto/src/main/java/net/consensys/cava/crypto/sodium/Signature.java]: You're right - I'll fix that later.

**Pull request merged**